File: 0785-is-graph-bipartite/0785-is-graph-bipartite.ts
function isBipartite(graph: number[][]): boolean {
    const n = graph.length
    const colors: number[] = Array(n).fill(0)
    for (let i=0; i < n; i++) {
        if (colors[i] === 0 && !dfs(i, 1)) return false
    }
    return true
    
    function dfs(i: number, color: number) {
        if (colors[i] !== 0) return colors[i] === color
        colors[i] = color
        for (const next of graph[i]) {
            if (!dfs(next, -color)) return false
        }
        return true
    }
};

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.py
class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        def dfs(nodes, i, color):
            if nodes[i] != 0 and nodes[i] != color:
                return False
            if nodes[i] == color:
                return True
            nodes[i] = color
            for next_node in graph[i]:
                if not dfs(nodes, next_node, -color):
                    return False
            return True
            
        n = len(graph)
        nodes = [0 for _ in range(0, n)]
        for i in range(0, n):
            if nodes[i] == 0:
                if not dfs(nodes, i, 1):
                    return False
        return True